### DISCUSSION TITLE
AssertJ 4

### DISCUSSION THREAD

**Discussion by scordio** [assertj/assertj#2229] [Ideas]
Let's define the scope of the next major version:

- [ ] #3442
- [ ] #3446
- [ ] Delete all [deprecated elements](https://github.com/assertj/assertj/search?q=deprecated&unscoped_q=deprecated)
- [ ] Internal error message refactoring
- [ ] use the new recursive comparison API for iterables element by element comparison, 
      this would lead to remove API made irrelevant like: 
  - [x] `usingComparatorForElementFieldsWithNames` 
  - [x] `usingComparatorForElementFieldsWithType`
  - [x] `usingElementComparatorOnFields`
- [ ] #1958
- [ ] Replace all the concrete return types on assertions / assumptions / soft assertions entry points with the corresponding `Abstract...Assert` type to facilitate extension for third-party libraries (similar to #2173, but using abstract classes instead)
- [ ] #2380
- [ ] Add [`Module`](https://docs.oracle.com/javase/9/docs/api/java/lang/Module.html) assertions and `module()` navigation method to `Class` assertions

Any comments, proposals or feedback?

**Comment by ishanarora04** [comment]
Hey , I would like to start work on few of these elements . Is it still available for grabs ?

**Comment by scordio** [comment]
@ishanarora04 sure it is! We would need a separate main branch to start merging version 4 topics, but this can be prepared once the first pull request is raised.

**Comment by bjhargrave** [comment]
>     * Drop multi-release build in favor of module-only

Does this mean JPMS modules only? We need to continue to support OSGi users. So the [osgi-test](https://github.com/osgi/osgi-test) team would like to work with you to ensure continued OSGi support.

**Comment by scordio** [comment]
@bjhargrave yes, that point was about JPMS. Working together making sure we are still supporting OSGi users properly would be perfect!

**Comment by onacit** [comment]
I'm not sure of the schedule but shouldn't Java-8 which is LTS-ish be still count?
Say, I'm the one who is pleased using Assert-J with [spring-boot-test](https://search.maven.org/search?q=a:spring-boot-test) which depends itself on assertj-*.
Shouldn't we consider those major frameworks/platforms regarding their plans for Java-11?
Thanks.

**Comment by scordio** [comment]
There is no schedule yet for AssertJ 4. Version 3 would easily continue along with version 4 for some time and we would backport any Java 8 compatible feature during this overlapping.

**Comment by scordio** [comment]
As per team's discussion, AssertJ 4 should use/require Java 17.

**Reply by bjhargrave** [reply]
> As per team's discussion, AssertJ 4 should use/require Java 17.

This is great. I assume that AssertJ 3 will also be maintained/updated for some time as well since many projects/users needing AssertJ goodness will need to support Java 8 for some years to come.

**Reply by scordio** [reply]
Yes, AssertJ 3 will stay around (long enough, I suppose) and get all Java 8 compatible features.

**Reply by scordio** [reply]
We are discussing with the team whether to stay on Java 8 or upgrade to Java 11 instead of upgrading to Java 17.

The main reason is to keep our effort reasonable and continue providing support to users who cannot upgrade yet (e.g., all RHEL 7 users under [ELS](https://www.redhat.com/en/blog/announcing-4-years-extended-life-cycle-support-els-red-hat-enterprise-linux-7)).

If we target Java 11, AssertJ 3 will enter maintenance mode and not receive new features.

Is there any strong reason why we should keep upgrading to Java 17?

**Reply by bjhargrave** [reply]
My $0.02. Continue to target Java 17 for AssertJ 4. Leave AssertJ 3 in maintenance mode for Java 8/11 users. Java 17 is already "old". Java 11 is super old. Java 17 is more interesting than Java 11 as an upgrade target for a project. Several projects I have worked on have gone from Java 8 straight to 17 since 17 had some compelling language features. Java 11 was not attractive enough to stimulate a move from Java 8.

**Reply by scordio** [reply]
If the users have a choice, I'm 100% with you.

However, there are legitimate enterprise cases where users cannot move away from Java 11. For example, in the RHEL 7 case that I mentioned:

> ELS for RHEL 7 is now available for 4 years, starting on July 1, 2024.

This means enterprises now have four more years to consider moving away from such a platform. From my point of view, this means that many projects might continue staying at most on Java 11, the maximum that RHEL 7 offers, for at least two more years.

We don't know yet if such a user base should pay the price of being left behind.

**Reply by scordio** [reply]
On a related note:

> Java 11 was not attractive enough to stimulate a move from Java 8.

That's also what we're considering if we don't go to Java 17. Would going to Java 11 have a benefit for the AssertJ internals? At least, I can think of some syntax sugars around optional and collections, plus https://github.com/orgs/assertj/discussions/2229#discussioncomment-783340). Whether these would be game changers, it needs to be evaluated.

**Reply by scordio** [reply]
We've received a lot of input on this topic from different sources. Thank you all!

As a result, we've decided that version 4 will require Java 17, and version 3 will go into maintenance mode after 3.27.1.

**Reply by ascopes** [reply]
What sort of timeframe would this be out of curiosity? I may as well baseline my library that uses AssertJ to Java 17 at the same time

**Reply by scordio** [reply]
We will release 3.27.1 latest at the beginning of January, and then we plan to start working on version 4.

We don't have a clear timeline for the first GA version but we definitely want to release intermediate milestones for early feedback, since we have in mind some major changes.

Any early integration would be greatly appreciated 🙂

**Reply by ascopes** [reply]
Sure thing. Give me a shout as soon as a release is available, I can get a branch set up to build from.

**Reply by ascopes** [reply]
I've got a v5 branch of my library up that targets a minimum version of Java 17. All I will need is either a snapshot repository (assertj could utilise github packages for this probably) or some form of distribution so I can build against it once you are ready.

**Reply by scordio** [reply]
We'll probably publish milestone and release candidate releases, as we do for [`assertj-assertions-generator`](https://repo.maven.apache.org/maven2/org/assertj/assertj-assertions-generator/).

**Reply by erizzo** [reply]
I see there was a first milestone release back in March, 2025 (8 months ago as of the time of this message). I totally understand that a major release can take a lot of time and most of the maintainers are probably working for free or in spare time. But it does make me wonder, has AssertJ 4 bitten off more than can be chewed, so to speak? Should the scope be reduced so that it can actually be delivered? I'd hate for it to turn into another [Duke Nukem Forever](https://www.guinnessworldrecords.com/world-records/most-protracted-game-development).

**Reply by scordio** [reply]
> has AssertJ 4 bitten off more than can be chewed, so to speak? Should the scope be reduced so that it can actually be delivered?

There are indeed major topics we'd like to address in version 4, and they do require some effort.

While putting them aside for now would indeed help speed up GA, we would miss the current "breaking change train", which is unlikely to happen again anytime soon (i.e., we don't want to release version 5 a year later). However, I totally understand the point, and we'll do our best to keep things to a minimum, especially for new features that could be easily added on top without a significant redesign of the core parts.

Lastly, just to update this thread, snapshots are now published to Maven Central.

**Reply by ascopes** [reply]
are there marked issues for deliverables that could be worked on in parallel by the community?

**Reply by scordio** [reply]
Currently, most of our effort is focused on reducing [the number of open PRs](https://github.com/assertj/assertj/pulls?q=is%3Aopen+is%3Apr) and preparing the migration guide [from 3.x to 4.0](https://github.com/assertj/assertj/wiki/AssertJ-4.0-Migration-Guide-(preview)).

We'll make sure to mark with [status: ideal for contribution](https://github.com/assertj/assertj/issues?q=state%3Aopen%20label%3A%22status%3A%20ideal%20for%20contribution%22) all the topics where the community could jump in!

**Comment by jojopad** [comment]
How about creating an extension for AssertJ that can be used as a general purpose fluent validation API? I think AssertJ's core API is well suited for such usage.

**Reply by scordio** [reply]
I'm not sure I understand what you mean. Would you mind adding some details?

**Comment by ascopes** [comment]
regarding #2639, you'd also want to update the implementation here to support ClassLoader#getName, etc. which were added in JRE 9!

If there is an active branch for AssertJ 4, then I can always cherrypick my changes onto that branch with the new functionality once I complete it?

**Reply by scordio** [reply]
AssertJ 4 development has not been started yet, but we plan to start working on it soon.

**Comment by onacit** [comment]
If I may, I want to put another comment.

Can we rebase the whole codebase based on interfaces for clean separation between specifications and implementations? Thanks.

**Reply by joel-costigliola** [reply]
That's something we want to do indeed.
Do you have specific interface candidates in mind you wish you had?

**Reply by onacit** [reply]
@joel-costigliola Thank.

I'm not concerning about *using* the AssertJ but about *extending* the AssertJ.
No matter how many times I tried to defined a clean interface extending `Assert<?, T>`, I keep myself falling back to `AbstractAssert<?, T>`.

I understand there were, are, and will be plenty of reasons of
* Using those` Abstract...Assert` classes as 1st class citizen
  * for, for an instance, (package-private|protected) variables and methods.
* Defining `NumberAssert<SELF, ACTUAL extends Number>` or `ComparableAssert<SELF, ATUAL extends Comparable<? super ACTUAL>>` as *not* extending `Assert<SELF, ACTUAL>`.
  * They are not *extended* types but traits.

When we're going to make the AssertJ4 based on Java 17 which clearly supports JPMS, we can construct a clean type/interface hierarchy and use those abstract/default classes as internal implementations.  

```java
interface Assert<SELF extends Assert<SELF, ACTUAL>, ACTUAL> {
        // empty
}

interface ObjectAssert<SELF extends Assert<SELF, ACTUAL>, ACTUAL>
        extends Assert<SELF, ACTUAL> {
}

interface NumberAssert<
            SELF extends NumberAssert<SELF, ACTUAL>,
            ACTUAL extends Number>
        extends ObjectAssert<SELF, ACTUAL> {
}

interface ComparableAssert<
            SELF extends ComparableAssert<SELF, ACTUAL>,
            ACTUAL extends Comparable<? super ACTUAL>>
        extends ObjectAssert<SELF, ACTUAL> {
}

interface ComparableAssert<
            SELF extends ComparableAssert<SELF, ACTUAL>,
            ACTUAL extends Comparable<? super ACTUAL>>
        extends ObjectAssert<SELF, ACTUAL> {
}

// trait
interface FloatingPointAssertTrait<
            ASSERT extends Assert<ASSERT> & FloatingPointNumberTrait<ASSERT, ACTUAL>,
            ACTUAL extends Number>
}

// and trait helper
final class FloatingPointAssertTraitHelper {
        // reflections and (method|var) handles
}

public interface FloatAssert<SELF extends FloatAssert<SELF>>
        extends NumberAssert<SELF, Float>,
                      ComparableAssert<SELF, Float>,
                      FloatingPointAssertTrait<SELF, Float> {
}

public abstract class AbstractFloatAssert<SELF extends AbstractFloatAssert<SELF>>
        extends AbstractNumberAssert<SELF, Float>
        implements FloatAssert<SELF> {
}

/* module private */ class FloatAsertImpl
        extends AbstractFloatAssert<FloatAssertImpl> {
}
```

The points are

* Let developers, who use AssertJ, rely on interfaces
* Let developers, who extends AssertJ, use those extensible abstract classes
* Hide default implementations which may not helping those using nor extending AssertJ. 

```java
final float actual = .0f;

// interface not abstract|concrete class
static FloatAssert assertThat(Float actual);

// extensible
static <ASSERT extends FloatAssert> assertThat(Float actual, AssertFactory<? super Float, ? extends ASSERT> factory);
```

**Reply by scordio** [reply]
> When we're going to make the AssertJ4 based on Java 17 which clearly supports JPMS, we can construct a clean type/interface hierarchy and use those abstract/default classes as internal implementations.

If we wouldn't expose such abstract classes and we would have interfaces only, maybe with default or even private methods, keeping the state needed for the type-related assertions would be a burden we push to the implementors of such interfaces. Or do you have something else in mind?

As a related topic, we might check how private methods in interfaces can help with it but I still feel we cannot have a stateless implementation.

Something that could be clearer from a user experience perspective, also for extension authors, is to drop the `Abstract` prefix in the class names and also reduce the visibility of the concrete classes to package-private, giving a clear indication that assertion instances should be obtained via entry point methods in `Assertions` or `BDDAssertions`.

> Hide default implementations which may not helping those using nor extending AssertJ.

I am very much in favor of this. Whenever possible, we should keep these things only with private or package-private visibility.

**Reply by scordio** [reply]
> * Let developers, who use AssertJ, rely on interfaces
> * Let developers, who extends AssertJ, use those extensible abstract classes

To continue satisfying the second point, we have to keep the abstract classes as `public`.

What users from the first point would gain if we would expose additional interfaces? I expect the common usage pattern to be a chain of assertions after an `assertThat` invocation so most likely there is no local variable holding the instance `assertThat` returns.

**Reply by onacit** [reply]
@scordio Thanks.

> To continue satisfying the second point, we have to keep the abstract classes as public.

Maybe yes or maybe no.

* We may include interfaces and abstract classes in a single module.
  * The point is that all AssertJ usages may be defined with only those interfaces.
  * Abstract classes are for those extending AssertJ. 
* We may separate interfaces and abstract classes into two separated modules.
   * Say, `assertj-core` and `assertj-runtime`?

Besides, I just did a quick sketch regarding interface separation.

* May looks somewhat off-road of traditional interface -> abstract class hierarachy.
* Depends on reflection/(method|var)handle

```java
public interface Assert<SELF extends Assert<SELF, ACTUAL>, ACTUAL> {

    default SELF isEqualTo(final Object expected) {
        // delegates operations to the delegate which commune with implementation class
        return AssertHelper.delegate((SELF) this).isEqualTo(expected);
    }
}

interface AssertDelegate<ASSERT extends Assert<ASSERT, ACTUAL>, ACTUAL> {

    ASSERT isEqualTo(final Object expected);
}

abstract class AbstractAssert<SELF extends AbstractAssert<SELF, DELEGATE, ACTUAL>, DELEGATE extends AssertDelegate<SELF, ACTUAL>, ACTUAL>
        implements Assert<SELF, ACTUAL> {

    protected AbstractAssert(final Class<SELF> clazz, final DELEGATE delegate, final ACTUAL actual) {
        super();
        this.clazz = Objects.requireNonNull(clazz, "clazz is null");
        this.actual = actual;
        this.delegate = Objects.requireNonNull(delegate, "delegate is null");
    }

    protected SELF self() {
        return clazz.cast(this);
    }

    private final Class<SELF> clazz;

    private final DELEGATE delegate;

    protected final ACTUAL actual;
}

abstract class AbstractAssertDelegate<ASSERT extends Assert<ASSERT, ACTUAL>, ACTUAL>
        implements AssertDelegate<ASSERT, ACTUAL> {

    AbstractAssertDelegate() {
        super();
    }

    @Override
    public ASSERT isEqualTo(final Object expected) {
        throw new UnsupportedOperationException("not implemented yet");
    }
}

final class AssertHelper {

    static <ASSERT extends Assert<ASSERT, ACTUAL>, ACTUAL> AssertDelegate<ASSERT, ACTUAL> delegate(final ASSERT assertion) {
        Objects.requireNonNull(assertion, "assertion is null");
        try {
            final Field field = AbstractAssert.class.getDeclaredField("delegate");
            field.setAccessible(true);
            return (AssertDelegate<ASSERT, ACTUAL>) field.get(assertion);
        } catch (final ReflectiveOperationException roe) {
            throw new RuntimeException("roe");
        }
    }

    private AssertHelper() {
        throw new AssertionError("instantiation is not allowed");
    }
}
```

**Comment by ascopes** [comment]
Is it likely that stuff such as `assertThatCode` will be deprecated in 4.x.x in favour of using `assertThatException` and `assertThatNoException` instead?

**Reply by scordio** [reply]
No, I don't think so. Once #2519 is finalized, `assertThatCode` should ideally become a superset of the other two, allowing the following:

```java
  @Test
  void demo() {
    Callable<Boolean> callable = () -> "value".isEmpty();

    // AbstractCallableAssert
    assertThatCode(callable).doesNotThrowAnyException()
                            .result() // returns AbstractObjectAssert
                            .isEqualTo(false);

    // AbstractCallableAssert
    assertThatCode(callable).result(BOOLEAN) // implicit doesNotThrowAnyException()
                            .isFalse();

    // AbstractCallableAssert
    assertThatCode(() -> "value".isEmpty()).doesNotThrowAnyException()
                                           .result()
                                           .isEqualTo(false);

    // AbstractCallableAssert
    assertThatCode(() -> nonVoidThrowingMethod()).hasThrownException() // returns ThrowableAssertAlternative
                                                 .withMessage("Boom!")
                                                 .withNoCause();

    ThrowableAssert.ThrowingCallable assertjThrowingCallable = () -> "value".isEmpty();

    // AbstractThrowableAssert
    assertThatCode(assertjThrowingCallable).doesNotThrowAnyException();

    // AbstractThrowableAssert
    assertThatCode(() -> System.out.println("")).doesNotThrowAnyException();

    // AbstractThrowableAssert
    assertThatCode(() -> voidThrowingMethod()).hasMessage("Boom!");
  }
```
